File: pufferlib/ocean/platformer/platformer.py
'''Simple Platformer Environment

A 2D platformer game where the player jumps between platforms to collect coins.
'''

import numpy as np
import gymnasium

import pufferlib
from pufferlib.ocean.platformer.cy_platformer import CyPlatformer


class Platformer(pufferlib.PufferEnv):
    def __init__(self, seed=None, num_envs=1, render_mode=None, report_interval=128,
             width=800, height=600, player_width=30, player_height=50, buf=None):

        # 12 observations: player (x,y,vx,vy) + 5 platforms (x,y)
        self.single_observation_space = gymnasium.spaces.Box(low=0, high=1,
            shape=(12,), dtype=np.float32)
        # 4 actions: left, right, jump, no action
        self.single_action_space = gymnasium.spaces.Discrete(4)
        self.report_interval = report_interval
        self.render_mode = render_mode
        self.num_agents = num_envs

        super().__init__(buf=buf)
        self.c_envs = CyPlatformer(self.observations, self.actions, self.rewards,
            self.terminals, num_envs, width, height, player_width, player_height)

    def reset(self, seed=None):
        self.c_envs.reset()
        self.tick = 0
        return self.observations, []

    def step(self, actions):
        self.actions[:] = actions
        self.c_envs.step()
        self.tick += 1

        info = []
        if self.tick % self.report_interval == 0:
            log = self.c_envs.log()
            if log['episode_length'] > 0:
                info.append(log)

        return (self.observations, self.rewards,
            self.terminals, self.truncations, info)

    def render(self):
        self.c_envs.render()

    def close(self):
        self.c_envs.close()


def test_performance(timeout=10, atn_cache=1024, num_envs=1024):
    import time

    env = Platformer(num_envs=num_envs)
    env.reset()
    tick = 0

    actions = np.random.randint(
        0,
        env.single_action_space.n,
        (atn_cache, num_envs),
    )

    start = time.time()
    while time.time() - start < timeout:
        atn = actions[tick % atn_cache]         
        env.step(atn)
        tick += 1

    print(f'SPS: {num_envs * tick / (time.time() - start)}')


if __name__ == '__main__':
    test_performance() 

File: pufferlib/ocean/platformer/platformer.c
#include "platformer.h"
#include "time.h"

void interactive() {
    CPlatformer env = {
        .width = 800,
        .height = 600,
        .player_width = 30,
        .player_height = 50,
    };
    allocate_cplatformer(&env);
    c_reset(&env);
 
    Client* client = make_client(env.width, env.height);
    
    int tick = 0;
    while (!WindowShouldClose()) {
        env.actions[0] = NOOP;  // Default to no action
        
        // Handle user input
        if (IsKeyDown(KEY_LEFT)) env.actions[0] = LEFT;
        if (IsKeyDown(KEY_RIGHT)) env.actions[0] = RIGHT;
        if (IsKeyPressed(KEY_SPACE)) env.actions[0] = JUMP;
        
        c_step(&env);
        c_render(client, &env);
    }
    
    close_client(client);
    free_allocated_cplatformer(&env);
}

void performance_test() {
    long test_time = 10;
    CPlatformer env = {
        .width = 800,
        .height = 600,
        .player_width = 30,
        .player_height = 50,
    };
    allocate_cplatformer(&env);
    c_reset(&env);
 
    long start = time(NULL);
    int i = 0;
    while (time(NULL) - start < test_time) {
        env.actions[0] = rand() % ACTIONS_COUNT;
        c_step(&env);
        i++;
    }
    long end = time(NULL);
    printf("SPS: %ld\n", i / (end - start));
    free_allocated_cplatformer(&env);
}

int main() {
    //performance_test();
    interactive();
    return 0;
} 